rfc: Session Authentication for HTTP Endpoints

mrtracy opened the pull request:
RFC which proposes session-based authentication for HTTP endpoints.

Commit:
rfc: Session Authentication for HTTP Endpoints

RFC which proposes session-based authentication for HTTP endpoints.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16829)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDk_kA1snLKe-JFwsg:-KoDk_kA1snLKe-JFwsh:bpwej5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L65)):*
> ```Markdown
> 
> ### Session Table
> A new metadata table will be created to hold system sessions:
> ```

This is a non-gossiped table, right? 

---

*[docs/RFCS/web_session_login.md, line 69 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDnJEEpBnGw9KUAPP4:-KoDnJEEpBnGw9KUAPP5:b-vm5vhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L69)):*
> ```Markdown
> ```sql
> CREATE TABLE system.sessions {
>     id          STRING      NOT NULL PRIMARY KEY,
> ```

How are session IDs generated? UUIDs? (We have a UUID type now that could be used here)

---

*[docs/RFCS/web_session_login.md, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDp6hVulurg-57vvr1:-KoDp6hVulurg-57vvr2:b-8snnty) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L71)):*
> ```Markdown
>     id          STRING      NOT NULL PRIMARY KEY,
>     username    STRING      NOT NULL,
>     created     TIMESTAMP   NOT NULL DEFAULT now(),
> ```

A coarse-grained "last used" timestamp could also be useful. We wouldn't want to update it every time but we could do so e.g. whenever the previous value was over an hour old.

---

*[docs/RFCS/web_session_login.md, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDkz_sWn5iUkZGOFkC:-KoDkz_sWn5iUkZGOFkD:b83qfk2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L74)):*
> ```Markdown
>     valid       BOOL        NOT NULL,
>     INDEX(created),
> }
> ```

For auditing, an additional free-form string field (containing JSON) will probably be useful. We should put the client's IP there to start and may come up with other metadata to include in the future. 

---

*[docs/RFCS/web_session_login.md, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDrdLVi-mCIUd5dYoE:-KoDrdLVi-mCIUd5dYoF:bsueln6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L109)):*
> ```Markdown
> 
> service Authentication {
>   rpc UserLogin(UserLoginRequest) returns (UserLoginResponse) {
> ```

There should also be a UserLogout endpoint to revoke the session and clear the cookie. (I guess it belongs outside the Authentication service since it needs valid cookies and CSRF protection)

---

*[docs/RFCS/web_session_login.md, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDoIKk9e88u1_HnC2J:-KoDoIKk9e88u1_HnC2K:boi1ohi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L132)):*
> ```Markdown
> appropriate HTTP response headers.
> 
> The "session" cookie is marked as "httponly" and is thus not accessible from
> ```

Both cookies should also have the `secure` attribute when secure mode is used. 

---

*[docs/RFCS/web_session_login.md, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDonYyJqPLA3GPiy-G:-KoDonYyJqPLA3GPiy-H:b8jl1ot) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L136)):*
> ```Markdown
> by XSS attacks.
> 
> If the "remember" flag is set to true on the request, the returned cookies
> ```

Is this useful these days? [Firefox persists session cookies across restarts](https://security.stackexchange.com/questions/31937/firefoxs-restore-previous-session-restores-logins-authenticated-by-servers-ses)

---

*[docs/RFCS/web_session_login.md, line 168 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDvKHC9_yqvRU9xMzz:-KoDvKHC9_yqvRU9xN--:b-a1t9f8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L168)):*
> ```Markdown
> ### CSRF Enforcement
> 
> CSRF is enforced using the "Double-submit cookie" method. This has three parts:
> ```

Can we combine this with validation of the Origin and Referer as described at [OWASP](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet), or does the potential for load balancers/proxies make it too tricky for us to determine the right origin to use? 

---

*[docs/RFCS/web_session_login.md, line 287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDtNEjXkNfGRbJOA3k:-KoDtNEjXkNfGRbJOA3l:b8kwefi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L287)):*
> ```Markdown
> on the token.
> 
> The major issue with JWT is that it does not provide a way to revoke login
> ```

Stateless sessions also introduce new issues around managing the keys to be used for signing and verifying sessions. 

---

*[docs/RFCS/web_session_login.md, line 313 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDulufCFaQd0cDA27a:-KoDulufCFaQd0cDA27b:buanbpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L313)):*
> ```Markdown
> If the CockroachDB server is running in insecure mode, we don't want to require
> user login on the frontend. However, the loading of page resources (html,
> javascript, etc.) is handled by embedded.go, and that does not give us an
> ```

Shouldn't the static resources in embedded.go simply be exempt from authentication in all cases? There's nothing secret there. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-:-KoDwZfeVAWAHFt8oge6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoFkwYiGDceG7Y_C3pd:-KoFkwYiGDceG7Y_C3pe:b54iq9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L35)):*
> any modifications made must be auditable

This seems like a good goal worthy of its own RFC, since AFAIK we aren't really making much auditable today.

---

*[docs/RFCS/web_session_login.md, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoFq7yghxJI5PfqFuht:-KoFq7yghxJI5PfqFuhu:b-vw33xr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L155)):*
> ```Markdown
> table. This is confirmed by performing a SELECT from the session table.
> 3. The corresponding session is valid (valid = true)
> 4. The age of its creation timestamp is less than system setting
> ```

It seems a little weird to me that previously expired sessions can become valid again if someone increases the setting. Is this standard practice (rather than also including an expiration time column)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-:-KoFrcbhgNaIeo3UPjpJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Are there any issues with the admin UI being served from multiple hostnames? What `domain` will be used for the cookie so that it can be sent by the browser to all of those hosts? Or will you have to log-in individually to each host you connect to?

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-:-KoI2d7ZoQSOlOV90j35:b-g4g29p)*
<!-- Sent from Reviewable.io -->

couchand commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIaYC78iElgMtGbTgi:-KoIaYC78iElgMtGbTgj:b9mqog5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L72)):*
> ```Markdown
>     username    STRING      NOT NULL,
>     created     TIMESTAMP   NOT NULL DEFAULT now(),
>     valid       BOOL        NOT NULL,
> ```

My general intuition is to replace boolean columns with a timestamp indicating when the event happened, in this case when the session was invalidated (and possibly by whom for auditability?).

---

*[docs/RFCS/web_session_login.md, line 146 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIX0gU6gyvQsbCRFaV:-KoIWl2-ojs5bHtmL299:b-llpvz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L146)):*
> ```Markdown
> only be added for the existing services; it will *not* be added for the new
> Authentication service, because the UserLogin method must be accessible without
> a session.
> ```

It is generally advisable to perform CSRF mitigation on the login endpoint; strictly validating the Referer header is likely the way to go there.

---

*[docs/RFCS/web_session_login.md, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIc0ylgVRMDRMAdB8h:-KoIc0ylgVRMDRMAdB8i:b4qf8e1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L159)):*
> ```Markdown
> 
> If any of the above criteria are not true, the incoming request is rejected with
> a 403 Forbidden message.
> ```

Should this rather be 401 Unauthorized?

---

*[docs/RFCS/web_session_login.md, line 170 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIXCLvY8fP779__9Se:-KoIXCLvY8fP779__9Sf:byh753t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L170)):*
> ```Markdown
> CSRF is enforced using the "Double-submit cookie" method. This has three parts:
> 
> + Upon user login (or any other successful request), a cryptographically-random
> ```

Does the parenthetical here mean that a particular CSRF token will only be valid for a single request?  It should be sufficient to have one token per session, which would minimize resource use - on the server, when generating the tokens, as well as a few bytes of network traffic, granted neither of those are that expensive.

---

*[docs/RFCS/web_session_login.md, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIZdyhx9XgpeCGloPi:-KoIZdyhx9XgpeCGloPj:b-cxgs4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L174)):*
> ```Markdown
> + When the client sends a request to the server, it reads the value from the
> cookie and adds it to an HTTP header "x-csrf-token".
> + When the server receives any message, it ensures that the value of the
> ```

Apparently this step is also not strictly necessary, as no current CSRF attack vector allows a forced request to add a custom header.  However, it does still seem prudent for defense in depth.

---

*[docs/RFCS/web_session_login.md, line 240 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIbgFa-r8usaEB4a7T:-KoIbgFa-r8usaEB4a7U:bb6j31r) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L240)):*
> ```Markdown
> + If the "logged in" value is present, the top-level "Layout" element will
> render the components of the currently requested route.
> + If any request comes back with a "403 Unauthorized", it is safe to assume that
> ```

401?

---

*[docs/RFCS/web_session_login.md, line 242 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIbUOBbSKloHC_oTVi:-KoIbUOBbSKloHC_oTVj:b-cm0r8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L242)):*
> ```Markdown
> + If any request comes back with a "403 Unauthorized", it is safe to assume that
> the user's session is no longer valid, and the "logged in" Admin UI state will
> thus be cleared.
> ```

It may be useful to maintain details of the failed request to retry after authentication.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-:-KoIdcbPlbPBY1f5PYvm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
@petermattis i'm under the impression that you will have to log in to each different hostname, they are different and sharing cookies would violate the same-origin policy.  There might be some kind of trick to get around it, but I don't know what it would be; probably something similar to how OAuth works, it would be complicated.

Addressed all the comments, this resulted in a number of modifications:

+ Structure of the "sessions" table has changed in a few ways
    + Timestamps have been modified to meet suggested criteria from comments
    + After offline discussion with Ben, I have added a new field "secret" which is used for security purposes, now explained in the RFC
    + Added "auditInfo" string field suggested by Ben

+ Added "Logout" method to Authentication service.
+ Changed CSRF a bit so that the "Login" method is now protected from CSRF.

---

Review status: 0 of 1 files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/web_session_login.md, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoFkwYiGDceG7Y_C3pd:-KokR-PvZ8oKPqU7D8z-:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L35)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> any modifications made must be auditable

This seems like a good goal worthy of its own RFC, since AFAIK we aren't really making much auditable today.
</blockquote></details>

:+1:

---

*[docs/RFCS/web_session_login.md, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDk_kA1snLKe-JFwsg:-KoeySbh5dM_GQ9bAY1H:bxpo0tl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L65)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This is a non-gossiped table, right? 
</blockquote></details>

Yes, and the requirement for a query of this table on every request is listed in the drawbacks system (along with its apparent solution, a short-term cache on each node).

---

*[docs/RFCS/web_session_login.md, line 69 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDnJEEpBnGw9KUAPP4:-KoezM0h1zFgf09KTTnQ:b-sflmjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L69)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

How are session IDs generated? UUIDs? (We have a UUID type now that could be used here)
</blockquote></details>

Yes, that will work.

The most important thing from a security standpoint is that the value is cryptographically random, so that a session ID can't be guessed from other random values taken from the server. Our UUIDs are generated using crypto/rand, so that is acceptable.

---

*[docs/RFCS/web_session_login.md, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDp6hVulurg-57vvr1:-Koi3VXrcbqrHpe3Wi64:b-kr3flo) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L71)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

A coarse-grained "last used" timestamp could also be useful. We wouldn't want to update it every time but we could do so e.g. whenever the previous value was over an hour old.
</blockquote></details>

I will go ahead and add this column based on your intuition, it could provide additional clues in the absence of auditing (and we may not want to audit read-only queries of all data sources).  If we don't use it can be removed (or ignored) later.

Added here, and added a note on when to update it in the "session enforcement" section.

---

*[docs/RFCS/web_session_login.md, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIaYC78iElgMtGbTgi:-Koi5oGix9megMG1oTPt:b-qry77w) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L72)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

My general intuition is to replace boolean columns with a timestamp indicating when the event happened, in this case when the session was invalidated (and possibly by whom for auditability?).
</blockquote></details>

I agree that "revokedAt" timestamp is likely more useful if an admin were reading a query on this table; changed.

As for the additional auditing information, i'm adding the free form "auditInfo" field suggested by ben below. This information could be recorded there.

---

*[docs/RFCS/web_session_login.md, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDkz_sWn5iUkZGOFkC:-Koi7eu-fs4rlTcxSqF7:bv8zi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L74)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

For auditing, an additional free-form string field (containing JSON) will probably be useful. We should put the client's IP there to start and may come up with other metadata to include in the future. 
</blockquote></details>

The IP may change during the course of a session; I would say that information should be included in audit logs, not in this table.

---

*[docs/RFCS/web_session_login.md, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDrdLVi-mCIUd5dYoE:-Koi7vSjbQNpU5LgUU9i:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L109)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

There should also be a UserLogout endpoint to revoke the session and clear the cookie. (I guess it belongs outside the Authentication service since it needs valid cookies and CSRF protection)
</blockquote></details>

Added.

---

*[docs/RFCS/web_session_login.md, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDoIKk9e88u1_HnC2J:-KoiUu4EQDpN_kn1iiZx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L132)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Both cookies should also have the `secure` attribute when secure mode is used. 
</blockquote></details>

Done.

---

*[docs/RFCS/web_session_login.md, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDonYyJqPLA3GPiy-G:-KoiVK7MBz_3FGU2Vqjj:b5640rh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L136)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is this useful these days? [Firefox persists session cookies across restarts](https://security.stackexchange.com/questions/31937/firefoxs-restore-previous-session-restores-logins-authenticated-by-servers-ses)
</blockquote></details>

As far as I can tell, that's basically up to the browser; apparently chrome will allow users to restore sessions even if the website specifically requested an expiration for the end of the session. https://stackoverflow.com/questions/17736737/cookie-to-expire-when-browser-session-ends

Doesn't seem to be anything we can really do about it, if the browser wants to keep sessions around even after you close the browser, that's their prerogative.

---

*[docs/RFCS/web_session_login.md, line 146 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIX0gU6gyvQsbCRFaV:-Kom9s__hp4sgWSlFaHw:b-97wl6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L146)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

It is generally advisable to perform CSRF mitigation on the login endpoint; strictly validating the Referer header is likely the way to go there.
</blockquote></details>

Well i'll say that the use case for a login CSRF for our product is difficult to understand. That said, i've changed things around; now all GRPC gateway methods are subject to CSRF, including login. The CSRF token is now set when static assets are retrieved on page load.

---

*[docs/RFCS/web_session_login.md, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoFq7yghxJI5PfqFuht:-KokR5rcrt7AUQMXBIta:b77xz2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L155)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It seems a little weird to me that previously expired sessions can become valid again if someone increases the setting. Is this standard practice (rather than also including an expiration time column)?
</blockquote></details>

you're right, I added an expiresAt field to the table.  The table description and this section have been updated.

---

*[docs/RFCS/web_session_login.md, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIc0ylgVRMDRMAdB8h:-KokImeUz9-6U-PvdH8x:buvwszt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L159)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Should this rather be 401 Unauthorized?
</blockquote></details>

So this is kind of odd.  If the client has a session cookie which has expired or has been revoked, 403 forbidden is semantically correct (as it is providing a session id as credentials, but that session is no longer valid).  If the client is *not* providing this token, then 401 is semantically correct.

However, 401 unauthorized necessarily returns the WWW-Authenticate header that instructs the browser how to authorize; common values include "Basic" and "Digest". These can cause the browser pop up a credentials dialog, which is something we don't want to do.

We could send back 401 with some sort of non-standard value for the WWW-authenticate field (i.e. "CustomLogin"), but i'm not sure what the behavior is supposed to be.  I'm having trouble finding best practices for this.

https://stackoverflow.com/questions/24553631/http-401-unauthorized-when-not-using-http-basic-auth

I don't think 403 is going to cause a problem, I'd prefer not to overthink this particular issue.

---

*[docs/RFCS/web_session_login.md, line 168 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDvKHC9_yqvRU9xMzz:-KokHlQ6KVAcp9-ylu3t:bqvsclc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L168)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can we combine this with validation of the Origin and Referer as described at [OWASP](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet), or does the potential for load balancers/proxies make it too tricky for us to determine the right origin to use? 
</blockquote></details>

We could, but I don't believe it's necessary for adequate security.

You are correct that it would a bit difficult; I think that the server has to have a list of valid origins, and this would likely have some sort of relation to the https certificate system.

I'm less confident about designing that particular system, so I'd like to leave it off for now as double-cookie provides adequate protection by itself.  We can add additional protection in the future if it becomes necessary, for example during a security audit.

---

*[docs/RFCS/web_session_login.md, line 170 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIXCLvY8fP779__9Se:-KokSiziHd1GByx1b-6h:b927hzb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L170)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Does the parenthetical here mean that a particular CSRF token will only be valid for a single request?  It should be sufficient to have one token per session, which would minimize resource use - on the server, when generating the tokens, as well as a few bytes of network traffic, granted neither of those are that expensive.
</blockquote></details>

No, the parenthetical was only meant to remark that the CSRF cookie could be set on any 200 OK request.

I've now changed the place where CSRF cookies are set in order to place the login method itself behind CSRF protection, so this has been removed.

---

*[docs/RFCS/web_session_login.md, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIZdyhx9XgpeCGloPi:-KokGFaEWSo3d3w-eZoc:b41rkab) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L174)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Apparently this step is also not strictly necessary, as no current CSRF attack vector allows a forced request to add a custom header.  However, it does still seem prudent for defense in depth.
</blockquote></details>

I don't think this is true - Seems like CSRF would be trivially easy if you just had to attach any header to the request.  What's the point of the cookie if the server doesn't need to validate the contents of x-csrf-token?

---

*[docs/RFCS/web_session_login.md, line 240 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIbgFa-r8usaEB4a7T:-KokEsHUF8ctwIB3KQ1z:b-otilcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L240)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

401?
</blockquote></details>

See comment above.

---

*[docs/RFCS/web_session_login.md, line 242 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIbUOBbSKloHC_oTVi:-KokFsZrTVWcDA5CV40p:btf0n7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L242)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

It may be useful to maintain details of the failed request to retry after authentication.
</blockquote></details>

Every action is dispatched based on a change in the redux state.  When the user logs back in, everything will resume as normal.

---

*[docs/RFCS/web_session_login.md, line 287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDtNEjXkNfGRbJOA3k:-KokFnA3yPKTWpRMCB9a:b-7jgylk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L287)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Stateless sessions also introduce new issues around managing the keys to be used for signing and verifying sessions. 
</blockquote></details>

Added a note.

---

*[docs/RFCS/web_session_login.md, line 313 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDulufCFaQd0cDA27a:-KokFP_lxeDxVMuQOlkW:bx2jren) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L313)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Shouldn't the static resources in embedded.go simply be exempt from authentication in all cases? There's nothing secret there. 
</blockquote></details>

Yes, they are. What this problem involves is that we don't want to pop up the login dialog if the server is in insecure mode; however, we can't tell if the server is insecure without making at least one request.

In any case, i've resolved this question (added solution to the "Insecure Mode" section, going to set a special cookie when returning index.html), and have removed this unresolved issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-:-KokPI9To4tYaEx4YsOI:b-rsmoez)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 1 files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/web_session_login.md, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDk_kA1snLKe-JFwsg:-KomgVzMvHY6-zt9c7wy:b-h7agy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L65)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Yes, and the requirement for a query of this table on every request is listed in the drawbacks system (along with its apparent solution, a short-term cache on each node).
</blockquote></details>

drawbacks *section

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 128 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KomtayRSBztzoYYQ2HS:-KomtayRSBztzoYYQ2HT:bmiee36) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fd8db68289d6e51170188d064fb8e902dca0db/docs/RFCS/web_session_login.md#L128)):*
> ```Markdown
> to the system.sessions table.
> 
> For other users, we provide a view over the table which has access to every
> ```

What's the expected use for this? I'm probably missing something, but it seems weird for un-privileged users to be able to view other users' sessions.

If we only wanted to expose the user's own sessions, I think we could do something like this if we were willing to implement the `SESSION_USER`, `CURRENT_USER`,  and/or `USER` functions:

```sql
CREATE VIEW system.my_sessions AS
SELECT id, username, createdAt, expiresAt, revokedAt, lastUsedAt, auditInfo
FROM system.sessions
WHERE username = SESSION_USER;
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 128 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KomtayRSBztzoYYQ2HS:-KonE3EydeT8LWL1AI9S:bp5u9xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fd8db68289d6e51170188d064fb8e902dca0db/docs/RFCS/web_session_login.md#L128)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

What's the expected use for this? I'm probably missing something, but it seems weird for un-privileged users to be able to view other users' sessions.

If we only wanted to expose the user's own sessions, I think we could do something like this if we were willing to implement the `SESSION_USER`, `CURRENT_USER`,  and/or `USER` functions:

```sql
CREATE VIEW system.my_sessions AS
SELECT id, username, createdAt, expiresAt, revokedAt, lastUsedAt, auditInfo
FROM system.sessions
WHERE username = SESSION_USER;
```
</blockquote></details>

The use case is for a non-root, administrator-like role.  For example, someone who has access to read audit logs, which presumably record the session id with each action. This view allows these non-root admins to view all session records without providing them enough information to impersonate those sessions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Made one additional change: the `id` column of the sessions table has been changed to type `SERIAL` instead of a UUID.  With the addition of the `secret` field, the id itself no longer needs to be cryptographically secure.  It also does not need to be globally unique, it only needs to be unique within the database itself. SERIAL is an INT, which will be smaller and more palatable to human readers (both helpful properties if this id is eventually used in auditing tables)

couchand commented:
Review status: 0 of 1 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 146 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIX0gU6gyvQsbCRFaV:-KorSHwPEino4mjQlABP:b-jei4ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L146)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Well i'll say that the use case for a login CSRF for our product is difficult to understand. That said, i've changed things around; now all GRPC gateway methods are subject to CSRF, including login. The CSRF token is now set when static assets are retrieved on page load.
</blockquote></details>

Honestly I agree that the attack vector for login CSRF is pretty hard to figure out, but since it's one of those things that's not hard to protect against it seems prudent.

---

*[docs/RFCS/web_session_login.md, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIc0ylgVRMDRMAdB8h:-KorTFacSm57VTEX3gG0:b-vxou7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L159)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

So this is kind of odd.  If the client has a session cookie which has expired or has been revoked, 403 forbidden is semantically correct (as it is providing a session id as credentials, but that session is no longer valid).  If the client is *not* providing this token, then 401 is semantically correct.

However, 401 unauthorized necessarily returns the WWW-Authenticate header that instructs the browser how to authorize; common values include "Basic" and "Digest". These can cause the browser pop up a credentials dialog, which is something we don't want to do.

We could send back 401 with some sort of non-standard value for the WWW-authenticate field (i.e. "CustomLogin"), but i'm not sure what the behavior is supposed to be.  I'm having trouble finding best practices for this.

https://stackoverflow.com/questions/24553631/http-401-unauthorized-when-not-using-http-basic-auth

I don't think 403 is going to cause a problem, I'd prefer not to overthink this particular issue.
</blockquote></details>

I'm not sure I agree that 403 is correct on an invalid session token.  From what I've seen 403 is meant as "you have a valid login but you can't see this resource" whereas 401 means "you're not logged in at all".  The SO link you've attached refers to Richardson and Amundsen's RESTful Web Services - here's the relevant page https://books.google.com/books?id=wWnGAAAAQBAJ&lpg=PA296&pg=PA306#v=onepage&q&f=false. A selected quote "It may have provided the wrong credentials, or none at all."  In my experience, 401 is a temporary condition that the client can reasonably expect to work out, but 403 is a permanent condition that says "don't bother trying again".

I agree that 403 won't "cause problems", but that argument could also be used to defend a stranger scheme such as HTTP 200 with a custom JSON payload `{ authenticated: false }` (which I've seen in practice, BTW).

---

*[docs/RFCS/web_session_login.md, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIZdyhx9XgpeCGloPi:-KorWXoufogAFCFYShqM:bmx5qhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L174)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I don't think this is true - Seems like CSRF would be trivially easy if you just had to attach any header to the request.  What's the point of the cookie if the server doesn't need to validate the contents of x-csrf-token?
</blockquote></details>

The double-submit mechanism is primarily intended to be used by passing the second copy of the value not as a custom header but as a hidden form field.

Yes, it's trivially easy to attach custom headers *if you control the entire request*.  But the threat model of CSRF is not that the attacker can send any arbitrary request (that's vacuously true), but that the attacker can force an unwitting user to send a request crafted by the attacker, which has very few handles to manipulate (of which custom headers are not one).

---

*[docs/RFCS/web_session_login.md, line 240 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIbgFa-r8usaEB4a7T:-KorXnrzpplxr39qTwDw:bbjxle1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L240)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

See comment above.
</blockquote></details>

In this case I suppose you mean 403 Forbidden, then, but see my comments.

---

*[docs/RFCS/web_session_login.md, line 242 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIbUOBbSKloHC_oTVi:-KorXcNaf3E9WNC1C0rh:bgcl1cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L242)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Every action is dispatched based on a change in the redux state.  When the user logs back in, everything will resume as normal.
</blockquote></details>

What about the initial request that received the authorization error?  It's not immediately clear to me how Redux will be responsible for retrying that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829)*
<!-- Sent from Reviewable.io -->

couchand commented:
Review status: 0 of 1 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIc0ylgVRMDRMAdB8h:-KorZ6TpMjFSgy4GWi_-:b-he6y35) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L159)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I'm not sure I agree that 403 is correct on an invalid session token.  From what I've seen 403 is meant as "you have a valid login but you can't see this resource" whereas 401 means "you're not logged in at all".  The SO link you've attached refers to Richardson and Amundsen's RESTful Web Services - here's the relevant page https://books.google.com/books?id=wWnGAAAAQBAJ&lpg=PA296&pg=PA306#v=onepage&q&f=false. A selected quote "It may have provided the wrong credentials, or none at all."  In my experience, 401 is a temporary condition that the client can reasonably expect to work out, but 403 is a permanent condition that says "don't bother trying again".

I agree that 403 won't "cause problems", but that argument could also be used to defend a stranger scheme such as HTTP 200 with a custom JSON payload `{ authenticated: false }` (which I've seen in practice, BTW).
</blockquote></details>

Actually getting making the distinction clear is important for the user experience, regardless of whether it's made through HTTP status codes or not.  Lack of authentication (which 401 communicates, despite the confusing name Unauthorized) means "redirect to the login page".  Lack of authorization (403 Forbidden) means "show some error page, cause you're not an admin"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDonYyJqPLA3GPiy-G:-KosRpCGkByoxUX9m2IO:b-b7vd3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L136)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

As far as I can tell, that's basically up to the browser; apparently chrome will allow users to restore sessions even if the website specifically requested an expiration for the end of the session. https://stackoverflow.com/questions/17736737/cookie-to-expire-when-browser-session-ends

Doesn't seem to be anything we can really do about it, if the browser wants to keep sessions around even after you close the browser, that's their prerogative.
</blockquote></details>

Right, there's not much we can do about it, but my point is that a session-lifetime cookie may live *longer* than a finite-lifetime cookie, so it's weird if checking "remember me" reduces the effective cookie lifetime. I don't think it makes sense to offer the checkbox at all. 

---

*[docs/RFCS/web_session_login.md, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIZdyhx9XgpeCGloPi:-KosSc9jweET93hj_UwI:b9u4lgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L174)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

The double-submit mechanism is primarily intended to be used by passing the second copy of the value not as a custom header but as a hidden form field.

Yes, it's trivially easy to attach custom headers *if you control the entire request*.  But the threat model of CSRF is not that the attacker can send any arbitrary request (that's vacuously true), but that the attacker can force an unwitting user to send a request crafted by the attacker, which has very few handles to manipulate (of which custom headers are not one).
</blockquote></details>

My understanding is that if a custom header is present, you know it's not CSRF, but custom headers cannot be added in all legitimate use cases (My information here is stale; I don't know if this is still the case). 

---

*[docs/RFCS/web_session_login.md, line 90 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KosRUWE4r_H5a-vzz0B:-KosRUWE4r_H5a-vzz0C:b-iwawvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c34402914a1f69ad92fc51acd3952b92c750bf24/docs/RFCS/web_session_login.md#L90)):*
> ```Markdown
> `secret` is a cryptographically random UUID generated and shared only with the
> original creator of the session. The server requires incoming requests to have
> the matching secret for their session key. This allows the session's key to be
> ```

s/key/id/g

---

*[docs/RFCS/web_session_login.md, line 309 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KosTA9z6FjGTGeOwiGo:-KosTA9z6FjGTGeOwiGp:b342slc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c34402914a1f69ad92fc51acd3952b92c750bf24/docs/RFCS/web_session_login.md#L309)):*
> ```Markdown
> + CSRF Token validation will not be applied.
> 
> Insecure mode is indicated to the client by setting a client-readable cookie
> ```

Why is a cookie needed for this, instead of inspecting `window.location` for `http` vs `https`?

Since we tend to run both `http` and `https` on the same port for different clusters, the `insecure` cookie could be persisted inappropriately between sessions (if developer runs an insecure cluster then switches to secure mode, for example). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829)*
<!-- Sent from Reviewable.io -->

couchand commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIZdyhx9XgpeCGloPi:-KosXV-reGp1bJiM_gnY:becn246) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L174)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

My understanding is that if a custom header is present, you know it's not CSRF, but custom headers cannot be added in all legitimate use cases (My information here is stale; I don't know if this is still the case). 
</blockquote></details>

I also had a question about how universal the custom header would be, particularly in enterprise environments where you are expecting a corporate man-in-the-middle.  Based on the results here https://seclab.stanford.edu/websec/csrf/csrf.pdf (almost 10 years old) it looks like it's a tiny fraction of real users that purge custom headers, so that seems like a reasonable way to go until/unless some client comes to us complaining that they can't access the Admin UI.  It seems like the advantage of the custom header is that it doesn't require a change to protobuf schemas to account for the CSRF token.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions.

---

*[docs/RFCS/web_session_login.md, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoDonYyJqPLA3GPiy-G:-KpI-Yj4dSohTpNjdNZs:bicbr9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L136)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Right, there's not much we can do about it, but my point is that a session-lifetime cookie may live *longer* than a finite-lifetime cookie, so it's weird if checking "remember me" reduces the effective cookie lifetime. I don't think it makes sense to offer the checkbox at all. 
</blockquote></details>

I'll remove it for now and we can add it in if it is requested later.

---

*[docs/RFCS/web_session_login.md, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIc0ylgVRMDRMAdB8h:-KpIEs5VJnTvgpBsnyBd:b-tn2f1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L159)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Actually getting making the distinction clear is important for the user experience, regardless of whether it's made through HTTP status codes or not.  Lack of authentication (which 401 communicates, despite the confusing name Unauthorized) means "redirect to the login page".  Lack of authorization (403 Forbidden) means "show some error page, cause you're not an admin"
</blockquote></details>

I am provisionally changing these to 401, but I have added an "unresolved question" at the end of the RFC.

If I can't get 401 to work without popping up the username/password dialog for basic auth, I will change them to 403 instead.

---

*[docs/RFCS/web_session_login.md, line 240 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIbgFa-r8usaEB4a7T:-KpII47OiSE-GslWDuS9:b-bang1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L240)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

In this case I suppose you mean 403 Forbidden, then, but see my comments.
</blockquote></details>

Answered above.

---

*[docs/RFCS/web_session_login.md, line 242 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIbUOBbSKloHC_oTVi:-KpI4a3bbycjVJosKruk:bc0p1am) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L242)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

What about the initial request that received the authorization error?  It's not immediately clear to me how Redux will be responsible for retrying that.
</blockquote></details>

Most of our requests are kicked off in response to a rendering event which finds out that the currently displayed data is out of date.  Once the user logs in, rendering the page will kick off the request again.

I think the only request that won't immediately be retried is persistently clearing one of the alerts; nothing else is currently modified by the user in the UI, it's read-only.

If we start making more advanced modifications, we might want to add some more advanced action handling at that point, but right now it's a bit of complication which is not necessary for the initial version of authentication.

---

*[docs/RFCS/web_session_login.md, line 90 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KosRUWE4r_H5a-vzz0B:-KpI-X8l0_P2wwnCzBWE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c34402914a1f69ad92fc51acd3952b92c750bf24/docs/RFCS/web_session_login.md#L90)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/key/id/g
</blockquote></details>

Done.

---

*[docs/RFCS/web_session_login.md, line 309 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KosTA9z6FjGTGeOwiGo:-KpI4SUjjOrYBTEuavN6:bcezltj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c34402914a1f69ad92fc51acd3952b92c750bf24/docs/RFCS/web_session_login.md#L309)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why is a cookie needed for this, instead of inspecting `window.location` for `http` vs `https`?

Since we tend to run both `http` and `https` on the same port for different clusters, the `insecure` cookie could be persisted inappropriately between sessions (if developer runs an insecure cluster then switches to secure mode, for example). 
</blockquote></details>

Yes, good insight, i've replaced the mechanism with a check of `window.location.protocol`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/web_session_login.md, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829#-KoIZdyhx9XgpeCGloPi:-KpIUjlJw4xm4kjZ6BmI:bss29qz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0e9cf54a19fb5c90a92808035761d0295bdb55c/docs/RFCS/web_session_login.md#L174)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I also had a question about how universal the custom header would be, particularly in enterprise environments where you are expecting a corporate man-in-the-middle.  Based on the results here https://seclab.stanford.edu/websec/csrf/csrf.pdf (almost 10 years old) it looks like it's a tiny fraction of real users that purge custom headers, so that seems like a reasonable way to go until/unless some client comes to us complaining that they can't access the Admin UI.  It seems like the advantage of the custom header is that it doesn't require a change to protobuf schemas to account for the CSRF token.
</blockquote></details>

I wasn't concerned with things stripping out custom headers, but with some browsers not allowing custom headers to be set at all. Looks like my information was very outdated, though: ["setRequestHeader was introduced in Windows Internet Explorer 7."](https://msdn.microsoft.com/en-us/library/ms536752(v=vs.85).aspx). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16829)*
<!-- Sent from Reviewable.io -->
